### PR TITLE
Ranked set additions for getting rank for values not present in the set

### LINKED ISSUE #425: RankedSet operation to return rank for values even when not present
Right now, a value missing from the set return `null`.

This alternative would return the position the value would occupy.

### LINKED ISSUE #426: Aggregate functions to return rank for values
Building on #425, these differ from the record functions that return its rank in that they take value(s) and the values might not be present in the index.

### PR DESCRIPTION
**Pull request by MMcM**
Resolves #425: RankedSet operation to return rank for values even when not present
Resolves #426: Aggregate functions to return rank for values

### CODE REVIEW

File: fdb-extensions/src/main/java/com/apple/foundationdb/async/RankedSet.java [hunk 1]
```diff
@@ -396,15 +405,30 @@ public long getRank() {
...
-            }));
+                containsCheckedKey(tr, key).thenCompose(exists -> {
+                    if (!exists && nullIfMissing) {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                    RankLookup rank = new RankLookup(key);
+                    return AsyncUtil.whileTrue(() -> nextLookup(rank, tr), executor)
+                            // If exists, rank is accurate. Else search ended at the next lower key, so are after it.
+                            // Except if key is before the leftmost key, in which case would take rank 0.
+                            .thenApply(vignore -> exists ? rank.getRank() : rank.wasBeforeFirst() ? 0 : rank.getRank() + 1);
```
[the remark this answers is not in the document]
Comment (MMcM) [reply fdb-extensions/src/main/java/com/apple/foundationdb/async/RankedSet.java]: I added another commit which moves the adjustment logic into `RankLookup`. This is less code and one fewer read. But I am not confident that it's actually clearer, even with a multi-line comment.

File: fdb-extensions/src/test/java/com/apple/foundationdb/async/RankedSetTest.java [hunk 2]
```diff
@@ -186,6 +186,22 @@ public void randomFiveThreads() throws InterruptedException {
         assertEquals("[]", uncaught.toString());
     }
 
+    @Test
+    public void rankAsThoughPresent() {
+        RankedSet rs = newRankedSet();
+        db.run(tr -> {
+            for (int i = 0; i < 100; i++) {
```
[the remark this answers is not in the document]
Comment (MMcM) [reply fdb-extensions/src/test/java/com/apple/foundationdb/async/RankedSetTest.java]: Sure. I had the two loops be parallel, but they are close enough even with this change.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/RankIndexMaintainer.java [hunk 3]
```diff
@@ -227,6 +228,16 @@ public boolean canEvaluateAggregateFunction(@Nonnull IndexAggregateFunction func
                     RankedSetIndexHelper.COMPARISON_SKIPPED_SCORE : null;
             return RankedSetIndexHelper.scoreForRank(state, rankedSet, (Number)values.get(groupingCount), outOfRange);
         }
+        if (FunctionNames.RANK_FOR_SCORE.equals(function.getName()) && range.isEquals()) {
+            Tuple values = range.getLow();
+            Subspace rankSubspace = getSecondarySubspace();
```
[the remark this answers is not in the document]
Comment (MMcM) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/RankIndexMaintainer.java]: It's easy enough with a functional style, which maybe isn't so bad.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/FunctionNames.java [hunk 4]
```diff
@@ -54,6 +54,10 @@
     public static final String SCORE_FOR_RANK_ELSE_SKIP = "score_for_rank_else_skip";
     public static final String SCORE_FOR_TIME_WINDOW_RANK_ELSE_SKIP = "score_for_time_window_rank_else_skip";
 
+    /* Rank for score functions */
+    public static final String RANK_FOR_SCORE = "rank_for_score";
```
[the remark this answers is not in the document]
Comment (MMcM) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/FunctionNames.java]: I amplified the comment a little for now.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/leaderboard/TimeWindowLeaderboardIndexMaintainer.java [hunk 5]
```diff
@@ -364,6 +370,48 @@ public boolean canEvaluateAggregateFunction(@Nonnull IndexAggregateFunction func
...
+            final Tuple groupKey = TupleHelpers.subTuple(tuple, 2, 2 + groupingCount);
+            final CompletableFuture<TimeWindowLeaderboard> leaderboardFuture = oldestLeaderboardMatching(type, timestamp);
+            return leaderboardFuture.thenCompose(leaderboard -> {
+                if (leaderboard == null) {
+                    return CompletableFuture.completedFuture(null);
+                }
+                final Tuple leaderboardGroupKey = leaderboard.getSubspaceKey().addAll(groupKey);
+                final Subspace extraSubspace = getSecondarySubspace();
+                final Subspace rankSubspace = extraSubspace.subspace(leaderboardGroupKey);
+                final RankedSet rankedSet = new RankedSetIndexHelper.InstrumentedRankedSet(state, rankSubspace, leaderboard.getNLevels());
```
[the remark this answers is not in the document]
Comment (MMcM) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/leaderboard/TimeWindowLeaderboardIndexMaintainer.java]: A regular function would need to asynchronously return three things, which seemed messier than having one that takes a callback. This also makes the rank and time-window leaderboard code much more similar, which seems desirable.

File: fdb-extensions/src/main/java/com/apple/foundationdb/async/RankedSet.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-extensions/src/main/java/com/apple/foundationdb/async/RankedSet.java]: Hm, I think it looks better, but I guess I could see it either way.

**Comment by MMcM** [comment]
I added a short release note. Let's just go with the latest version of the code.

**Pull request merged**